### ISSUE TITLE
Check BIP39 generates invalid SSKR shares

### ISSUE THREAD

**Issue by InfiniteQE** [aido/app-seed-tool#23]
1. Reset Ledger Nano S Plus
2. Restore mnemonic seed (all all all all all all all all all all all all)
3. Install app-seed-tool 1.5.4
4. Run Seed Tool
5. Check BIP39 (all all all all all all all all all all all all)
6. Generate SSKR Shares, 2of3
```
tuna acid epic gyro brew diet able acid able high yell very atom whiz paid king legs door yell junk city ruby pose gush stub able able taco liar
tuna acid epic gyro brew diet able acid acid cook zaps rust iron tuna body draw race lava pose easy gray diet into nail item able able oval lung
tuna acid epic gyro brew diet able acid also trip wave omit trip onyx safe toil webs inch fuel when luck main dark toil memo able able part real
```

7. Enter any 2 of these into BlockchainCommon's `seedtool --in sskr` on MacOS

```
seedtool --in sskr                                                           
tuna acid epic gyro brew diet able acid able high yell very atom whiz paid king legs door yell junk city ruby pose gush stub able able taco liar
tuna acid epic gyro brew diet able acid acid cook zaps rust iron tuna body draw race lava pose easy gray diet into nail item able able oval lung
seedtool: Invalid Bytewords.
```
```
seedtool --in sskr
tuna acid epic gyro brew diet able acid also trip wave omit trip onyx safe toil webs inch fuel when luck main dark toil memo able able part real
tuna acid epic gyro brew diet able acid able high yell very atom whiz paid king legs door yell junk city ruby pose gush stub able able taco liar
seedtool: Invalid Bytewords.
```
```
seedtool --in sskr
tuna acid epic gyro brew diet able acid able high yell very atom whiz paid king legs door yell junk city ruby pose gush stub able able taco liar
tuna acid epic gyro brew diet able acid also trip wave omit trip onyx safe toil webs inch fuel when luck main dark toil memo able able part real
seedtool: Invalid Bytewords.
```

**Comment by aido** [comment]
Hi @InfiniteQE ,

I performed some tests on a Ledger S for BIP39 phrase `all all all all all all all all all all all all` and the following SSKR shares were generated:
```
SSKR share 1:
tuna acid epic gyro undo pose able acid able keep hang ramp horn gray half bias heat part maze data wall zaps wand jugs dice soap eyes knob judo
SSKR share 2:
tuna acid epic gyro undo pose able acid acid inky keep hawk play jugs whiz cost logo open join ramp echo king bias kiwi scar love love idle axis
SSKR share 3:
tuna acid epic gyro undo pose able acid also guru brew junk wave body arch dull view puma jump cats frog vial fair iron view obey poem fact away
```

I then tested these phrases with the seedtool CLI and all looks good:
```
$ seedtool --in bip39 "all all all all all all all all all all all all"
0660cc198330660cc198330660cc1983
$ seedtool --in sskr
tuna acid epic gyro undo pose able acid able keep hang ramp horn gray half bias heat part maze data wall zaps wand jugs dice soap eyes knob judo
tuna acid epic gyro undo pose able acid acid inky keep hawk play jugs whiz cost logo open join ramp echo king bias kiwi scar love love idle axis
0660cc198330660cc198330660cc1983
$ seedtool --in sskr
tuna acid epic gyro undo pose able acid able keep hang ramp horn gray half bias heat part maze data wall zaps wand jugs dice soap eyes knob judo
tuna acid epic gyro undo pose able acid also guru brew junk wave body arch dull view puma jump cats frog vial fair iron view obey poem fact away
0660cc198330660cc198330660cc1983
$ seedtool --in sskr
tuna acid epic gyro undo pose able acid acid inky keep hawk play jugs whiz cost logo open join ramp echo king bias kiwi scar love love idle axis
tuna acid epic gyro undo pose able acid also guru brew junk wave body arch dull view puma jump cats frog vial fair iron view obey poem fact away
0660cc198330660cc198330660cc1983
```

I will test with other devices but the code for generating shares is common to all devices.

> [!NOTE]
> I am using Speculos rather than a physical device for testing

**Comment by aido** [comment]
Hi again @InfiniteQE,

I have now also tested on a Nano S Plus.  Test was a success with valid SSKR shares generated,
The following SSKR shares were generated by the Nano S Plus:
```
SSKR share 1:
tuna acid epic gyro game free able acid able free days liar user data next chef fund film knob guru flap atom cyan beta atom ramp play love chef
SSKR share 2:
tuna acid epic gyro game free able acid acid crux dull data gush flew film navy kick fish hard runs gush oboe yurt lazy vows lava beta jowl kiwi
SSKR share 3:
tuna acid epic gyro game free able acid also many edge unit tied warm rust aqua door fern easy math jade gyro luck bald safe saga crux omit holy
```
I then tested these phrases with the seedtool CLI and all looks good:
```
$ seedtool --in bip39 "all all all all all all all all all all all all"
0660cc198330660cc198330660cc1983
$ seedtool --in sskr
tuna acid epic gyro game free able acid able free days liar user data next chef fund film knob guru flap atom cyan beta atom ramp play love chef
tuna acid epic gyro game free able acid acid crux dull data gush flew film navy kick fish hard runs gush oboe yurt lazy vows lava beta jowl kiwi
0660cc198330660cc198330660cc1983
$ seedtool --in sskr
tuna acid epic gyro game free able acid able free days liar user data next chef fund film knob guru flap atom cyan beta atom ramp play love chef
tuna acid epic gyro game free able acid also many edge unit tied warm rust aqua door fern easy math jade gyro luck bald safe saga crux omit holy
0660cc198330660cc198330660cc1983
$ seedtool --in sskr
tuna acid epic gyro game free able acid acid crux dull data gush flew film navy kick fish hard runs gush oboe yurt lazy vows lava beta jowl kiwi
tuna acid epic gyro game free able acid also many edge unit tied warm rust aqua door fern easy math jade gyro luck bald safe saga crux omit holy
0660cc198330660cc198330660cc1983
```

May I ask you to please double check your own test procedure, thanks.


> [!NOTE]
> Again, I am using Speculos rather than a physical device for testing

**Comment by InfiniteQE** [comment]
Nano S Plus

1. Wipe
2. Restore Seed : `beef beef beef beef beef beef beef beef beef beef beef beef`
3. Ledger Live > Install > Recovery Check
4. MacOS Terminal, ~Downloads/compiled_app_binaries.v1_5_4/  
```
python -m ledgerblue.loadApp --appFlags 0x10 --tlv --targetId 0x33100004 --targetVersion="" --apiLevel 1 --delete --fileName nanos2/bin/app.hex --appName "Seed Tool" --appVersion "1.5.4" --dataSize $((0x`cat debug/app.map |grep _envram_data | tr -s ' ' | cut -f2 -d' '|cut -f2 -d'x'` - 0x`cat debug/app.map |grep _nvram_data | tr -s ' ' | cut -f2 -d' '|cut -f2 -d'x'`)) `ICONHEX=\`python3 /opt/nanosplus-secure-sdk/icon3.py --hexbitmaponly glyphs/seed_nanox.gif  2>/dev/null\` ; [ ! -z "$ICONHEX" ] && echo "--icon $ICONHEX"`  --curve secp256k1 --path ""
```
5. Recovery Check > 12 words > `beef beef beef beef beef beef beef beef beef beef beef beef` > Phrase is correct
6. Seed Tool > Check BIP39 recovery phrase > 12 words > `beef beef beef beef beef beef beef beef beef beef beef beef` > BIP39 Phrase is correct > Generate SSKR phrases > Select number of shares > 3 > Select threshold > 2 > `<            >` (no words, just left right arrows on screen edges, frozen app)
7. restart Ledger
8. Seed Tool > Check BIP39 recovery phrase > 12 words > `beef beef beef beef beef beef beef beef beef beef beef beef` > BIP39 Phrase is correct > pressing both buttons on this screen, exits the app
9. Seed Tool > Check BIP39 recovery phrase > 12 words > `beef beef beef beef beef beef beef beef beef beef beef beef` > BIP39 Phrase is correct > Generate SSKR phrases > Select number of shares > 3 > Select threshold > 2 > 
```
tuna acid epic gyro view knob able acid able scar void zero numb days rock brew jade away hard vial hard heat user down fact able able cook gush
tuna acid epic gyro view knob able acid acid many view away buzz claw miss poem beta liar cusp lion loud crux when trip gems able able luck solo
tuna acid epic gyro view knob able acid also help vial atom tent high vial jury obey bald plus diet very pool legs tomb tomb able able peck wasp
```
10. MacOS
```
seedtool --in sskr
tuna acid epic gyro view knob able acid able scar void zero numb days rock brew jade away hard vial hard heat user down fact able able cook gush
tuna acid epic gyro view knob able acid acid many view away buzz claw miss poem beta liar cusp lion loud crux when trip gems able able luck solo
^D
seedtool: Invalid Bytewords.

seedtool --in sskr
tuna acid epic gyro view knob able acid able scar void zero numb days rock brew jade away hard vial hard heat user down fact able able cook gush
tuna acid epic gyro view knob able acid also help vial atom tent high vial jury obey bald plus diet very pool legs tomb tomb able able peck wasp
^D
seedtool: Invalid Bytewords.

seedtool --in sskr
tuna acid epic gyro view knob able acid acid many view away buzz claw miss poem beta liar cusp lion loud crux when trip gems able able luck solo
tuna acid epic gyro view knob able acid also help vial atom tent high vial jury obey bald plus diet very pool legs tomb tomb able able peck wasp
^D
seedtool: Invalid Bytewords.
```

11. Using your Speculos example above 
```
seedtool --in sskr
tuna acid epic gyro game free able acid able free days liar user data next chef fund film knob guru flap atom cyan beta atom ramp play love chef
tuna acid epic gyro game free able acid acid crux dull data gush flew film navy kick fish hard runs gush oboe yurt lazy vows lava beta jowl kiwi
^D
0660cc198330660cc198330660cc1983

seedtool --in sskr
tuna acid epic gyro game free able acid acid crux dull data gush flew film navy kick fish hard runs gush oboe yurt lazy vows lava beta jowl kiwi
tuna acid epic gyro game free able acid also many edge unit tied warm rust aqua door fern easy math jade gyro luck bald safe saga crux omit holy
^D
0660cc198330660cc198330660cc1983

seedtool --in sskr
tuna acid epic gyro game free able acid able free days liar user data next chef fund film knob guru flap atom cyan beta atom ramp play love chef
tuna acid epic gyro game free able acid also many edge unit tied warm rust aqua door fern easy math jade gyro luck bald safe saga crux omit holy
^D
0660cc198330660cc198330660cc1983
```

12. MacOS Seedtool test
```
seedtool --in bip39 "beef beef beef beef beef beef beef beef beef beef beef beef"
140280500a0140280500a0140280500a

seedtool --in hex --out sskr --group 2-of-3 140280500a0140280500a0140280500a
tuna acid epic gyro solo idle able acid able ugly buzz join love peck oval leaf lion inky brew many tent yell idle fern taco solo luck half zoom
tuna acid epic gyro solo idle able acid acid plus navy many game limp unit kick acid pool wasp mint knob drum kiln fizz kiwi part knob note yurt
tuna acid epic gyro solo idle able acid also epic blue purr body work good jowl menu wave yawn ruby meow holy hard sets quad menu next peck oboe

seedtool --in sskr                                                          
tuna acid epic gyro solo idle able acid able ugly buzz join love peck oval leaf lion inky brew many tent yell idle fern taco solo luck half zoom
tuna acid epic gyro solo idle able acid acid plus navy many game limp unit kick acid pool wasp mint knob drum kiln fizz kiwi part knob note yurt
^D
140280500a0140280500a0140280500a

seedtool --in sskr
tuna acid epic gyro solo idle able acid acid plus navy many game limp unit kick acid pool wasp mint knob drum kiln fizz kiwi part knob note yurt
tuna acid epic gyro solo idle able acid also epic blue purr body work good jowl menu wave yawn ruby meow holy hard sets quad menu next peck oboe
^D
140280500a0140280500a0140280500a
```

**Comment by aido** [comment]
Hi @InfiniteQE ,

So in your tests above you performed the exact same actions in step 6:

> 6. Seed Tool > Check BIP39 recovery phrase > 12 words > `beef beef beef beef beef beef beef beef beef beef beef beef` > BIP39 Phrase is correct > Generate SSKR phrases > Select number of shares > 3 > Select threshold > 2 > `<            >` (no words, just left right arrows on screen edges, frozen app)

as you did for step 9:
> 9. Seed Tool > Check BIP39 recovery phrase > 12 words > `beef beef beef beef beef beef beef beef beef beef beef beef` > BIP39 Phrase is correct > Generate SSKR phrases > Select number of shares > 3 > Select threshold > 2 >
tuna acid epic gyro view knob able acid able scar void zero numb days rock brew jade away hard vial hard heat user down fact able able cook gush
tuna acid epic gyro view knob able acid acid many view away buzz claw miss poem beta liar cusp lion loud crux when trip gems able able luck solo
tuna acid epic gyro view knob able acid also help vial atom tent high vial jury obey bald plus diet very pool legs tomb tomb able able peck wasp

and had 2 different outcomes?

First time (step 6) the app froze and with the same actions a second time (step 9) the app generates some shares.

**Comment by InfiniteQE** [comment]
correct

**Comment by aido** [comment]
That does not make sense to me. I am baffled.

**Comment by aido** [comment]
Hi @InfiniteQE,

Would it be correct to sum up the issue as follows:

* When testing on Speculos there is no issue, valid shares are generated which can be validated by the seedtool CLI also
* When testing on a physical Ledger S Plus device sometimes the device freezes and sometimes the device generates invalid SSKR shares

The inconsistent behaviour of the second point is quite strange,

**Comment by InfiniteQE** [comment]
Correct, as of the release binaries of 1.5.4 on a physical Nano S Plus, I haven't tested any other builds or code branches this round of testing, app-seed-tool always generates invalid shards.

The device freeze issue has only happened once.

**Comment by aido** [comment]
Hi @InfiniteQE,

As I do not have a Ledger Nano S Plus device this is a difficult bug for me to reproduce, All tests I perform on Speculos pass but the same tests reportedly fail on an actual physical device.

I had a similar issue last month with  how Ledger were implementing the `os_derive_bip32_no_throw(()` function in their  [app-recovery-check](https://github.com/LedgerHQ/app-recovery-check) application. The app-seed-tool application inherited that issue from  [app-recovery-check](https://github.com/LedgerHQ/app-recovery-check) and it had very similar symptoms to what you are reporting here i.e. works on Speculos but not on device. That particular issue was fixed here: https://github.com/LedgerHQ/app-recovery-check/commit/f3e486d5787d3a6324b1061fdc339cd06f948262.
@lpascal-ledger found and fixed that particular problem quite quickly and it was subsequently fixed in app-seed-tool by me.

This leads me to suspect that the issue you are reporting lies with the implementation of one of the Ledger functions and the difference between real firmware and emulated firmware i.e. this is a Ledger issue not an app issue. However, pinpointing where or in what function the issue lies is difficult.

Looking at your detailed troubleshooting steps above the issue happens just after the user choses the threshold for the shares, This will hopefully help me hone in on the problematic function

**Comment by aido** [comment]
I have narrowed the possible cause of this issue down to a list of five suspects: `os_swap_u32()`, `os_secure_memcmp()`, `cx_crc32_hw()`, `cx_rng()` or `cx_hmac_sha256()`.
I suspect one of these functions is doing something different on a physical device to what it is doing in Speculos,

**Comment by InfiniteQE** [comment]
Nano S Plus

1. Wipe
2. Restore Seed : `trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade`
3. Ledger Live > Install > Recovery Check
4. MacOS Terminal, re-download compiled_app_binaries.v1_5_4.zip from Github, extract to a new directory.  Confirm all /nanos2/bin files have the same MD5 as the version downloaded 2 days ago.
5. ~/Downloads/compiled_app_binaries.v1_5_4/  
```
python -m ledgerblue.loadApp --appFlags 0x10 --tlv --targetId 0x33100004 --targetVersion="" --apiLevel 1 --delete --fileName nanos2/bin/app.hex --appName "Seed Tool" --appVersion "1.5.4" --dataSize 0 --icon 0100000000ffffff00ffffffffffffffc3f1c1c3c083e107fb0fff3ffe7ffe7ffe7ffe7ffeffffffff --curve secp256k1 --path ""
```
6. Seed Tool > Check BIP39 recovery phrase > 24 words > `trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade` > BIP39 Phrase is correct > Generate SSKR phrases > Select number of shares > 3 > Select threshold > 2 > 
```
tuna acid epic hard data drop draw able acid able toil love keep idle city down meow epic drop zoom pose ruby void obey redo jowl warm luau jazz join brew what jury data urge race cook visa hawk arch huts hope able able cost girl

tuna acid epic hard data drop draw able acid acid list quiz iced kick logo judo mild horn slot inch drum vial rich urge drop oval junk view hope skew kept hope buzz lava cusp plus zest whiz cash days heat wand able able meow gray

tuna acid epic hard data drop draw able acid also jugs yawn help holy data next mint note veto taxi ramp acid help hope loud visa rust hang back crux ugly race ramp jury figs list safe taco silk flux cash drum able able high judo
```
7. *note another bug, when using the Right button to scroll to the last page of the last share (3 shares, each 4 pages), the next screen is Quit.  When reversing back from Quit to the beginning with the Left button, after the 5th press the backwards sequence starts jumping to page 1 (skipping pages 2,3,4) and sometimes jumps straight to Quit.  Also advancing beyond Quit with either repeated Left or repeated Right button presses also jumps around the shares pages incorrectly.*
8. MacOS
```
seedtool --in sskr                                                                            
tuna acid epic hard data drop draw able acid able toil love keep idle city down meow epic drop zoom pose ruby void obey redo jowl warm luau jazz join brew what jury data urge race cook visa hawk arch huts hope able able cost girl
tuna acid epic hard data drop draw able acid acid list quiz iced kick logo judo mild horn slot inch drum vial rich urge drop oval junk view hope skew kept hope buzz lava cusp plus zest whiz cash days heat wand able able meow gray
^D
seedtool: Invalid Bytewords.

seedtool --in sskr
tuna acid epic hard data drop draw able acid able toil love keep idle city down meow epic drop zoom pose ruby void obey redo jowl warm luau jazz join brew what jury data urge race cook visa hawk arch huts hope able able cost girl
tuna acid epic hard data drop draw able acid also jugs yawn help holy data next mint note veto taxi ramp acid help hope loud visa rust hang back crux ugly race ramp jury figs list safe taco silk flux cash drum able able high judo
^D
seedtool: Invalid Bytewords.

seedtool --in sskr
tuna acid epic hard data drop draw able acid acid list quiz iced kick logo judo mild horn slot inch drum vial rich urge drop oval junk view hope skew kept hope buzz lava cusp plus zest whiz cash days heat wand able able meow gray
tuna acid epic hard data drop draw able acid also jugs yawn help holy data next mint note veto taxi ramp acid help hope loud visa rust hang back crux ugly race ramp jury figs list safe taco silk flux cash drum able able high judo
^D
seedtool: Invalid Bytewords.
```

9. MacOS
```
seedtool --in bip39 "trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade"
e6bcd79af35e6bcd79af35e6bcd79af35e6bcd79af35e6bcd79af35e6bcd79af

seedtool --in hex --out sskr --group 2-of-3 e6bcd79af35e6bcd79af35e6bcd79af35e6bcd79af35e6bcd79af35e6bcd79af 
tuna acid epic hard data open good able acid able time poem veto loud aunt when arch acid hard inky part hang ruin aqua inch monk vows oval undo leaf code flew jury also next frog when menu poem kick void solo pose high yell junk
tuna acid epic hard data open good able acid acid solo item yoga keno saga item flap grim wall hawk horn runs lion exit idle puff gush wall claw wolf surf curl buzz holy aqua taco solo undo days zinc calm join trip fund ruby webs
tuna acid epic hard data open good able acid also void exit tomb jazz mint undo lung owls curl cats crux news taco kite into soap luau exit gala horn owls liar ramp road quiz gift lazy belt numb item cook navy urge obey taxi gala

seedtool --in sskr   
tuna acid epic hard data open good able acid able time poem veto loud aunt when arch acid hard inky part hang ruin aqua inch monk vows oval undo leaf code flew jury also next frog when menu poem kick void solo pose high yell junk
tuna acid epic hard data open good able acid acid solo item yoga keno saga item flap grim wall hawk horn runs lion exit idle puff gush wall claw wolf surf curl buzz holy aqua taco solo undo days zinc calm join trip fund ruby webs
e6bcd79af35e6bcd79af35e6bcd79af35e6bcd79af35e6bcd79af35e6bcd79af

seedtool --in hex --out bip39 e6bcd79af35e6bcd79af35e6bcd79af35e6bcd79af35e6bcd79af35e6bcd79af
trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade trade
```

**Comment by aido** [comment]
Hi @InfiniteQE ,

The last 4 words of every SSKR phrase is the checksum. I find it strange that in all the examples you provided that were generated on a physical device the first 2 words of every checksum is always `able able` i.e. 0x0000. This is not the case with examples generated by Speculos. The issue may be with how the `cx_crc32_hw()` function is working on a physical device. I will investigate further ... getting closer.

**Comment by aido** [comment]
Oh no ... I just remembered. I think I may have written the CRC32 implementation that is being used by Speculos :-)

https://github.com/LedgerHQ/speculos/issues/363


@xchapron-ledger  suggested here: https://github.com/LedgerHQ/speculos/issues/363#issuecomment-1491609979 that Ledger's implementation of CRC32 in Speculos was broken, so I fixed it. I wonder was it broken in the real firmware too? 

This is a bold statement to make but I think that the implementation of CRC32 in the Ledger firmware (not Speculos) may be broken.

**Comment by xchapron-ledger** [comment]
@aido According to this: https://github.com/LedgerHQ/ledger-secure-sdk/blob/master/lib_cxng/src/cx_crc32.c#L24, the crc32_hw implementation is the CRC-32-IEEE802.3. I'm afraid you implemented the CRC32-CCITT32 in Speculos?

**Comment by xchapron-ledger** [comment]
Seem's like CRC32-CCITT32 is the same thing as CRC-32-IEEE802.3. Would be great to check if a simple call to `cx_crc32()` give the same results on Speculos and on device to conclude...

**Comment by xchapron-ledger** [comment]
I just tested it, And I confirm that this is not the same result.

**Comment by aido** [comment]
The CRC16 implementation used in Speculos is CRC-16/CCITT. And now, after my "fix", the CRC32 implementation in Speculos is CRC-32/CCITT.

**Comment by aido** [comment]
> I just tested it, And I confirm that this is not the same result.

So, either the Speculos implementation is broken or the firmware implementation is broken?

Please please be the firmware, I wrote the Speculos version!!! :-)

Seriously though, the Speculos version of `cx_crc32_hw()` seems to work when we are testing this app but the firmware version does not.

This tool may be handy for testing both versions. 
https://crccalc.com/

Also note the following lines in the Speculos version before I "fixed" it:

https://github.com/LedgerHQ/speculos/blob/3cf2a3cde7a42605fe1a7b82f6f6c04424259181/src/bolos/cxlib.c#L29-L37

Also, Is this possibly a big endian versus little endian issue?

**Comment by lpascal-ledger** [comment]
This issue is currently been investigated on the OS side, we'll come back to you shortly (by the end of the week max.).

**Comment by aido** [comment]
Nice.... 

https://github.com/LedgerHQ/ledger-secure-sdk/pull/500


I am curious to know if the Nano S has the same issue?

**Comment by lpascal-ledger** [comment]
Yes, so, as you understood it is indeed buggy in the OS. Including NanoS which, being EOL, should not receive an update just for this.
So short term OS calls will be bypassed as it's going to be implemented directly into the SDK, then partially re-routed to OS calls depending on NanoS+ / NanoX / Stax fix updates.

Anyway, sorry for the undue sweat, and thanks again for using these rarely-called functions! That blows some dust off these old bones.

**Comment by aido** [comment]
> Including Nano S which, being EOL, should not receive an update just for this.

That's what I feared,

Great that the issue was found but for now I may just roll my own implementation of CRC32 into the app-seed-tool app so that it works on all devices, including the EOL Nano S.

**Comment by xchapron-ledger** [comment]
NanoS will be patched on next SDK update. But the patch will stay on SDK side as probably not Os released will be done.

**Comment by aido** [comment]
>  as probably not OS released will be done.


https://github.com/LedgerHQ/ledger-app-builder/blob/d6ca0fe11b518fc208d71222989784380330ef68/lite/Dockerfile#L68C1-L69C54

Was hoping this line meant another Nano S OS would be released in the future. But maybe not as the device is EOL

```
# Latest Nano S SDK
# Will switch to the unified SDK for next OS release.
```

**Comment by aido** [comment]
Hi,

I have added a `cx_crc32()` function to the app as follows:

```C
uint32_t cx_crc32(const uint8_t *data, size_t len) {
    uint32_t crc = ~0;
    const uint8_t *end = data + len;

    while (data < end) {
        crc ^= *data++;
        for (uint8_t i = 0; i < 8; i++) {
            uint32_t mask = ~((crc & 1) - 1);
            crc = (crc >> 1) ^ (0xEDB88320 & mask);
        }
    }
    return ~crc;
}
```

This is a temporary workaround. When Ledger have fixed `cx_crc32_hw()` for the devices this app needs to run on then I will remove `cx_crc32()`  and start using  `cx_crc32_hw()` again.

app-seed-tool v1.6.0 is now released:
https://github.com/aido/app-seed-tool/releases/tag/v1.6.0

**Comment by InfiniteQE** [comment]
It would be worth keeping that function, as there will be some users who don't want to upgrade firmware

**Comment by aido** [comment]
Issue fixed in 9a3041ba6e96c7858fdaafb35761efb8715e9250

**Comment by xchapron-ledger** [comment]
@aido :wave: 
We just deployed a new version of the SDK fixing the cx_crc32 implementation for all device OS in production (this includes Nanos 2.1.0).
I think you should remove your version from the app code in order to benefit from it (and automatically of the future in HW).

I also don't recommend shadowing cx function like that, as IMO there is no way to be sure of the version that will be selected by the linker script?
I also don't recommend prefixing your functions with `cx_`, this is misleading as their are not really part of the cx lib as other `cx_` prefixed functions..
